ql/lib/codeql/kaleidoscope/Cfg.qll

```diff
@@ -0,0 +1,254 @@
+    override string toString() { result = "successor" }
+  }
+
+  class BooleanSuccessor extends SuccessorType, TBooleanSuccessor {
+    boolean value;
+
+    BooleanSuccessor() { this = TBooleanSuccessor(value) }
+
+    override string toString() { result = value.toString() }
+
+    boolean getValue() { result = value }
+  }
+
+  class ReturnSuccessor extends SuccessorType, TReturnSuccessor {
+    override string toString() { result = "return" }
+  }
+}
+
+private module CfgScope {
+  abstract class CfgScope extends AstNode { }
+
+  private class FunctionScope extends CfgScope, FunctionDefinition { }
+
+  private class ProgramScope extends CfgScope, Program { }
+}
+
+private module Implementation implements CfgShared::InputSig<Location> {
+  import codeql.kaleidoscope.Ast
+  import Completion
+  import CfgScope
+
+  predicate completionIsNormal(Completion c) { not c instanceof ReturnCompletion }
+
+  // Not using CFG splitting, so the following are just a dummy types.
+  private newtype TUnit = Unit()
+
+  class SplitKindBase = TUnit;
+
+  class Split extends TUnit {
+    abstract string toString();
+  }
+
+  predicate completionIsSimple(Completion c) { c instanceof SimpleCompletion }
+
+  predicate completionIsValidFor(Completion c, AstNode e) { c.isValidFor(e) }
+
+  CfgScope getCfgScope(AstNode e) {
+    exists(AstNode p | p = e.getParent() |
+      result = p
+      or
+      not p instanceof CfgScope and result = getCfgScope(p)
+    )
+  }
+
+  int maxSplits() { result = 0 }
+
+  predicate scopeFirst(CfgScope scope, AstNode e) {
+    first(scope.(Program), e) or
+    first(scope.(FunctionDefinition).getBody(), e)
+  }
+
+  predicate scopeLast(CfgScope scope, AstNode e, Completion c) {
+    last(scope.(Program), e, c) or
+    last(scope.(FunctionDefinition).getBody(), e, c)
+  }
+
+  predicate successorTypeIsSimple(SuccessorType t) { t instanceof NormalSuccessor }
+
+  predicate successorTypeIsCondition(SuccessorType t) { t instanceof BooleanSuccessor }
+
+  SuccessorType getAMatchingSuccessorType(Completion c) { result = c.getAMatchingSuccessorType() }
+
+  predicate isAbnormalExitType(SuccessorType t) { none() }
+}
+
+import CfgShared::Make<Location, Implementation>
+import Completion
+import CfgScope
+
+private class ProgramTree extends StandardPreOrderTree instanceof Program {
+  override ControlFlowTree getChildNode(int i) { result = super.getStatement(i) }
+}
+
+private class FunctionDefinitionTree extends LeafTree instanceof FunctionDefinition { }
+
+private class FunctionCallExpressionTree extends StandardPostOrderTree instanceof FunctionCallExpression
+{
+  override ControlFlowTree getChildNode(int i) { result = super.getArgument(i) }
+}
+
+private class BinaryOpExpressionTree extends StandardPostOrderTree instanceof BinaryOpExpression {
+  override ControlFlowTree getChildNode(int i) {
+    result = super.getLhs() and i = 0
+    or
+    result = super.getRhs() and i = 1
```

Comment:
Again, what about short-curcuiting?

Reply:
There is not such thing in kaleidoscope ;-)